materialize-mongodb/driver.go

```diff
@@ -0,0 +1,201 @@
+  if err := req.Validate(); err != nil {
+    return nil, fmt.Errorf("validating request: %w", err)
+  }
+
+  es := schemagen.GenerateSchema("Materialize MongoDB Spec", &config{})
+  endpointSchema, err := es.MarshalJSON()
+  if err != nil {
+    return nil, fmt.Errorf("generating endpoint schema: %w", err)
+  }
+
+  resourceSchema, err := schemagen.GenerateSchema("MongoDB Collection", &resource{}).MarshalJSON()
+  if err != nil {
+    return nil, fmt.Errorf("generating resource schema: %w", err)
+  }
+
+  return &pm.SpecResponse{
+    EndpointSpecSchemaJson: json.RawMessage(endpointSchema),
+    ResourceSpecSchemaJson: json.RawMessage(resourceSchema),
+    DocumentationUrl:       "https://go.estuary.dev/materialize-mongodb",
+    Oauth2Spec:             nil,
+  }, nil
+}
+
+func (d driver) Validate(ctx context.Context, req *pm.ValidateRequest) (*pm.ValidateResponse, error) {
+  cfg, err := resolveEndpointConfig(req.EndpointSpecJson)
+  if err != nil {
+    return nil, err
+  }
+
+  _, err = d.Connect(ctx, cfg)
+  if err != nil {
+    return nil, fmt.Errorf("connecting to database: %w", err)
+  }
+
+  // MongoDB supports arbitrary JSON documents so there are no constraints on
+  // any fields and all fields are marked as "recommended"
+  var out []*pm.ValidateResponse_Binding
+  for _, b := range req.Bindings {
+    res, err := resolveResourceConfig(b.ResourceSpecJson)
+    if err != nil {
+      return nil, err
+    }
+
+    constraints := make(map[string]*pm.Constraint)
+    for _, projection := range b.Collection.Projections {
+      var constraint = new(pm.Constraint)
+      constraint.Type = pm.Constraint_LOCATION_RECOMMENDED
+      constraint.Reason = "JSON fields are supported and included by default"
+      constraints[projection.Field] = constraint
+    }
+
+    resourcePath := []string{res.Database, res.Collection}
+
+    out = append(out, &pm.ValidateResponse_Binding{
+      Constraints:  constraints,
+      DeltaUpdates: false,
+      ResourcePath: resourcePath,
+    })
+  }
+
+  return &pm.ValidateResponse{Bindings: out}, nil
+}
+
+// ApplyUpsert is no-op for MongoDB since collections are created on the fly. We
+// only check for connection to database.
+func (d driver) ApplyUpsert(ctx context.Context, req *pm.ApplyRequest) (*pm.ApplyResponse, error) {
+  if err := req.Validate(); err != nil {
+    return nil, fmt.Errorf("validating request: %w", err)
+  }
+
+  cfg, err := resolveEndpointConfig(req.Materialization.EndpointSpecJson)
+  if err != nil {
+    return nil, err
+  }
+
+  _, err = d.Connect(ctx, cfg)
+  if err != nil {
+    return nil, fmt.Errorf("connecting to database: %w", err)
+  }
+
+  return &pm.ApplyResponse{
+    ActionDescription: "",
+  }, nil
+}
+
+// ApplyDelete is no-op as well, since this operation is going to be superseded
+// by an "Apply" operation with no bindings specified
+func (d driver) ApplyDelete(ctx context.Context, req *pm.ApplyRequest) (*pm.ApplyResponse, error) {
```

Comment:
Unless this connector is using the JSON protocol it will still get `ApplyDelete` requests, right? And if it were using the JSON protcol, my understand is that `ApplyUpsert` would need to handle the "Apply with no bindings" to actually delete the collections in the endpoint.

Is the intent here to simply leave the endpoint collections alone on delete? That wouldn't seem totally unreasonable to me, although it would be different than (most?) other materialization connectors we have which do perform some action on the endpoint when a materialization is deleted from flow.

Reply:
I think the idea for removing `ApplyDelete` and making it an `ApplyUpsert` with no bindings is exactly that: to avoid deleting things when a materialization is deleted.

My understanding is, if you run a materialization to a destination, and then decide to remove the materialization from Estuary dashboard, you should be able to keep your data.